src/tls/ffi.c

```diff
@@ -0,0 +1,402 @@
+        state->fstar_state = 0;
+        free(state);
+    }
+}
+
+void FFI_mitls_free_msg(char *msg)
+{
+    free(msg);
+}
+
+void FFI_mitls_free_packet(void *packet)
+{
+    free(packet);
+}
+
+void * copypacket(value packet, /* out */ size_t *packet_size)
+{
+    void *p;
+    mlsize_t size;
+        
+    size = caml_string_length(packet);
+    p = malloc(size);
+    if (p) {
+        memcpy(p, String_val(packet), size);
+        *packet_size = size;
+    }
+    return p;
+}
+
+// Called from FStar code to send via TCP
+CAMLprim value ocaml_send_tcp(value cookie, value bytes)
+{
+    mlsize_t buffer_size;
+    char *buffer;
+    int retval;
+    size_t c = Long_val(cookie);
+    struct _FFI_mitls_callbacks *callbacks = (struct _FFI_mitls_callbacks *)c;
+    CAMLparam2(cookie, bytes);
+
+    buffer = Bp_val(bytes);
+    buffer_size = caml_string_length(bytes);
+    
+    retval = (*callbacks->send)(callbacks, buffer, buffer_size);
+    
+    return Val_int(retval);
+}
+
+// Called from FStar code to receive via TCP
+CAMLprim value ocaml_recv_tcp(value cookie, value bytes)
+{
+    mlsize_t buffer_size;
+    char *buffer;
+    ssize_t retval;
+    size_t c = Long_val(cookie);
+    struct _FFI_mitls_callbacks *callbacks = (struct _FFI_mitls_callbacks *)c;
+    CAMLparam2(cookie, bytes);
+    
+    buffer = Bp_val(bytes);
+    buffer_size = caml_string_length(bytes);
+    
+    retval = (*callbacks->recv)(callbacks, buffer, buffer_size);
+    
+    return Val_int(retval);
+}
+
+// Called by the host app to create a TLS connection.
+int FFI_mitls_connect(struct _FFI_mitls_callbacks *callbacks, /* in */ mitls_state *state, /* out */ char **outmsg, /* out */ char **errmsg)
+{
+    CAMLlocal1(result);
+    int ret;
+    mitls_redirect r;
+    
+    *outmsg = NULL;
+    *errmsg = NULL;
+    redirect_stdio(&r);
+    
+    result = caml_callback2_exn(*state->connect_callback, state->fstar_state, Val_long((size_t)callbacks));
+    if (Is_exception_result(result)) {
+        printf("Exception!  %s\n", caml_format_exception(Extract_exception(result)));
+        ret = 0;
+    } else {
+        // The result is a closure, usable to encrypt and decrypt packets
+        state->fstar_state = result;
+        ret = 1;
+        
+    }
+    restore_stdio(&r, outmsg, errmsg);
+    return ret;
+}
+
+// Called by the host app to encode a cleartext packet for transmission
+void * FFI_mitls_prepare_send(/* in */ mitls_state *state, const void* buffer, size_t buffer_size, /* out */ size_t *packet_size, /* out */ char **outmsg, /* out */ char **errmsg)
+{
+    void *p = NULL;
+    CAMLlocal2(buffer_value, result);
+    mitls_redirect r;
+
+    *outmsg = NULL;
+    *errmsg = NULL;
+    redirect_stdio(&r);
+    
+    buffer_value = caml_alloc_string(buffer_size);
+    memcpy(Bp_val(buffer_value), buffer, buffer_size);
+    
+    result = caml_callback3_exn(state->fstar_state, Val_int(0), buffer_value, Val_unit);
```

Comment:
can you please define, on the C side: `enum { FFI_Encrypt = 0, FFI_Decrypt }`, then on the OCaml side: `type action = Encrypt | Decrypt`, then use these to pass or examine the first parameter of `state->fstar_state`?